### ISSUE TITLE
Using a learning-rate scheduler?

### ISSUE THREAD

**Issue by egillax** [dfalbel/torchoptx#1]
Do you think it would be a lot of work to get this working with a learning rate scheduler? I really like the speedup I'm getting with these optimizer's but to use it for my package I need to make it work with a learning rate scheduler. I see there are not many schedulers in libtorch and not the ones I want to use are not there. 

I did some experimenting locally and managed to extract the default learning rate from the cpp optimizer, I think I could use that to set the default learning rate.  But I think for it to work the scheduler classes in R you need to be able to loop over those in R like in the base step function:

```R
step = function() {
      self$last_epoch <- self$last_epoch + 1
      values <- self$get_lr()

      for (i in seq_along(self$optimizer$param_groups)) {
        self$optimizer$param_groups[[i]]$lr <- values[i]
        self$print_lr(self$verbose, i, self$optimizer$param_groups[[i]]$lr)
      }
```

I think this means you have to implement at least:
- Get optimizer param_groups with `$` 
- Access one optimizer param group with `[[i]]` and get it's learning rate with `$`, and use `$ <- ` to assign
- Be able to do `seq_along` on the param_groups, and `lapply` I guess.

Anything I'm forgetting?

**Comment by dfalbel** [comment]
I think for most schedulers that's all you need. This could be accomplished with:

1. In the c++ side implement accessors for querying the vector of parameter groups: https://pytorch.org/cppdocs/api/classtorch_1_1optim_1_1_optimizer.html#_CPPv4N5torch5optim9Optimizer12param_groupsEv
2. Then an accessor for indexing this vector.
3. Accessor for getting and setting the optimizer options: https://pytorch.org/cppdocs/api/classtorch_1_1optim_1_1_optimizer_param_group.html#_CPPv4N5torch5optim19OptimizerParamGroup7optionsEv
4. Implement S3 methods for `$`, `$<-`, `[[` and `[[<-` that call this accessors. 

It's not too much work, but requires a reasonable understanding of how torchexport, etc works. Hopefully, the same implementation can be easily used for all optimizers.

**Comment by egillax** [comment]
I made a first try to create an accessor for the parameter groups [here](https://github.com/egillax/torchoptx/tree/lr_scheduler)

But when I try to create the optimizer with the parameter groups I get an error: 

```Error in optim_param_group(opt) : 
  Expecting an external pointer: [type=raw].
```

Do you know what it means?
